<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/20-summa-cifr-faktoriala' style="text-decoration: none; color: #1a73e8;">
            20-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

## Задача 20

Факториал числа `n` определяется следующим образом:
$$
n! = n \cdot (n - 1) \cdot \ldots \cdot 3 \cdot 2 \cdot 1
$$

Например:
$$
10! = 10 \cdot 9 \cdot \ldots \cdot 3 \cdot 2 \cdot 1 = 3628800
$$

Сумма цифр числа `10!` равна:
$$
3 + 6 + 2 + 8 + 8 + 0 + 0 = 27
$$

Необходимо найти сумму цифр числа
$$
100!
$$

Ответ: `648`

## 1) Рекурсивное вычисление факториала

Сначала самостоятельно вычислим значение `100!` с помощью рекурсивной функции, а затем найдём сумму цифр полученного числа.

Для вычисления факториала используется классическое рекуррентное соотношение:

$$
n! =
\begin{cases}
1, & n = 1,\\
n \cdot (n-1)!, & n > 1.
\end{cases}
$$

После вычисления факториала сумма цифр находится тем же способом, который был рассмотрен в <a href='https://github.com/AlexeyM4/M4_algorithms_and_algorithmic_problem_solving/blob/main/Project%20Euler/Task_016.ipynb'>**16-й задаче проекта Эйлера**</a>: число преобразуется в строку, каждая цифра переводится в целое число, после чего все значения суммируются функцией `sum`.

---

### Алгоритм

1. Рекурсивно вычислить значение `100!`.
2. Преобразовать полученное число в строку.
3. Преобразовать каждую цифру в целое число.
4. Найти сумму всех цифр.
5. Вернуть полученный результат.

---

### Особенности

- факториал реализован самостоятельно с помощью рекурсии;
- для подсчёта суммы цифр используется подход из предыдущей задачи;
- код получается простым и наглядным.

---

### Сложность

Вычисление факториала требует `O(n)` умножений, где `n` — число, факториал которого вычисляется.

Подсчёт суммы цифр выполняется за `O(k)`, где `k` — количество цифр в числе `100!`.

Итоговая сложность:

- **Время:** `O(n + k)`.
- **Дополнительная память:** `O(n)` — из-за глубины рекурсии.

In [8]:
from time import perf_counter # измерение времени выполнения алгоритмов

In [9]:
def factorial_1(n: int) -> int:
    if n == 1:
        return 1
    return n * factorial_1(n - 1)


def get_result_1() -> int:
    return sum(map(int, str(factorial_1(100))))


start = perf_counter()
print(get_result_1())
end = perf_counter()

print(f'{end - start:.7f}')

648
0.0001908


## 2) Итеративное вычисление факториала

В отличие от предыдущего решения, факториал вычисляется с помощью цикла. Последовательно перемножаются все числа от `1` до `100`.

Сумма цифр полученного числа вычисляется тем же способом, который был рассмотрен в <a href='https://github.com/AlexeyM4/M4_algorithms_and_algorithmic_problem_solving/blob/main/Project%20Euler/Task_016.ipynb'>**16-й задаче проекта Эйлера**</a>.

---

### Алгоритм

1. Вычислить значение `100!` с помощью цикла.
2. Найти сумму цифр полученного числа.

---

### Особенности

- факториал вычисляется без рекурсии;
- отсутствуют накладные расходы на рекурсивные вызовы;
- код остаётся простым и эффективным.

---

### Сложность

- **Время:** `O(n + k)`, где `n` — число, факториал которого вычисляется, а `k` — количество цифр в результате.
- **Дополнительная память:** `O(1)`.

In [10]:
def factorial_2(n: int) -> int:
    f = 1
    for i in range(1, n + 1):
        f *= i
    return f


def get_result_2() -> int:
    return sum(map(int, str(factorial_2(100))))


start = perf_counter()
print(get_result_2())
end = perf_counter()

print(f'{end - start:.7f}')

648
0.0000973


## 3) Использование встроенной функции `factorial`

В этом решении для вычисления факториала используется готовая функция `factorial` из модуля `math`.

Сумма цифр вычисляется тем же способом, который был рассмотрен в <a href='https://github.com/AlexeyM4/M4_algorithms_and_algorithmic_problem_solving/blob/main/Project%20Euler/Task_016.ipynb'>**16-й задаче проекта Эйлера**</a>.

---

### Алгоритм

1. Вычислить значение `100!` с помощью функции `factorial`.
2. Найти сумму цифр полученного числа.

---

### Особенности

- используется готовая реализация из стандартной библиотеки;
- решение получается максимально компактным;
- код легко читается и не требует собственной реализации факториала.

---

### Сложность

- **Время:** `O(n + k)`.
- **Дополнительная память:** `O(k)`.

In [11]:
from math import factorial


def get_result_3() -> int:
    return sum(map(int, str(factorial(100))))


start = perf_counter()
print(get_result_3())
end = perf_counter()

print(f'{end - start:.7f}')

648
0.0001073


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **10_000 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [12]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
    get_result_3,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(), number=10_000) / 10_000
    print(f'Алгоритм {i}: {tm:.10f}')

Алгоритм 1: 0.0000192138
Алгоритм 2: 0.0000148147
Алгоритм 3: 0.0000090242


## Вывод

В данной задаче требовалось найти сумму цифр числа `100!`.

Были рассмотрены три способа вычисления факториала. Первый использует рекурсивную реализацию, второй — итеративный алгоритм с циклом, а третий — встроенную функцию `factorial` из стандартного модуля `math`. Во всех случаях сумма цифр вычисляется одинаковым способом, рассмотренным ранее в <a href='https://github.com/AlexeyM4/M4_algorithms_and_algorithmic_problem_solving/blob/main/Project%20Euler/Task_016.ipynb'>**16-й задаче проекта Эйлера**</a>.

Результаты измерения времени выполнения:

- **Алгоритм 1:** `0.0000192138` с;
- **Алгоритм 2:** `0.0000148147` с;
- **Алгоритм 3:** `0.0000090242` с.

Из результатов видно, что рекурсивная реализация уступает итеративной из-за накладных расходов на рекурсивные вызовы. Циклический вариант работает быстрее, однако наилучший результат показывает встроенная функция `factorial`, реализованная на уровне стандартной библиотеки Python и оптимизированная для работы с большими числами.

Таким образом, для практического использования наиболее предпочтительным является **третье решение**, так как оно сочетает простоту, компактность и максимальную производительность.